homeassistant/components/roon/media_player.py

```diff
@@ -0,0 +1,518 @@
+    DEVICE_DEFAULT_NAME,
+    STATE_IDLE,
+    STATE_OFF,
+    STATE_PAUSED,
+    STATE_PLAYING,
+)
+from homeassistant.core import callback
+from homeassistant.helpers.dispatcher import (
+    async_dispatcher_connect,
+    async_dispatcher_send,
+)
+from homeassistant.util import convert
+from homeassistant.util.dt import utcnow
+
+from .const import DOMAIN
+
+DEPENDENCIES = ["roon"]
+
+SUPPORT_ROON = (
+    SUPPORT_PAUSE
+    | SUPPORT_VOLUME_SET
+    | SUPPORT_STOP
+    | SUPPORT_PREVIOUS_TRACK
+    | SUPPORT_NEXT_TRACK
+    | SUPPORT_SHUFFLE_SET
+    | SUPPORT_SEEK
+    | SUPPORT_TURN_ON
+    | SUPPORT_TURN_OFF
+    | SUPPORT_VOLUME_MUTE
+    | SUPPORT_PLAY
+    | SUPPORT_PLAY_MEDIA
+    | SUPPORT_SELECT_SOURCE
+    | SUPPORT_VOLUME_STEP
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up Roon MediaPlayer from Config Entry."""
+    roon_server = hass.data[DOMAIN][config_entry.entry_id]
+    media_players = set()
+
+    @callback
+    def async_update_media_player(player_data):
+        """Add or update Roon MediaPlayer."""
+        dev_id = player_data["dev_id"]
+        if dev_id not in media_players:
+            # new player!
+            media_player = RoonDevice(roon_server, player_data)
+            media_players.add(dev_id)
+            async_add_entities([media_player])
+        else:
+            # update existing player
+            async_dispatcher_send(
+                hass, f"room_media_player_update_{dev_id}", player_data
+            )
+
+    # start listening for players to be added or changed by the server component
+    async_dispatcher_connect(hass, "roon_media_player", async_update_media_player)
+
+
+class RoonDevice(MediaPlayerEntity):
+    """Representation of an Roon device."""
+
+    def __init__(self, server, player_data):
+        """Initialize Roon device object."""
+        self._remove_signal_status = None
+        self._sources = []
+        self._server = server
+        self._available = True
+        self._last_position_update = None
+        self._supports_standby = False
+        self._state = STATE_IDLE
+        self._last_playlist = None
+        self._last_media = None
+        self._last_changed = None
+        self._unique_id = None
+        self._zone_id = None
+        self._output_id = None
+        self._name = DEVICE_DEFAULT_NAME
+        self._media_title = None
+        self._media_album_name = None
+        self._media_artist = None
+        self._media_position = 0
+        self._media_duration = 0
+        self._is_volume_muted = False
+        self._volume_step = 0
+        self._source = None
+        self._shuffle = False
+        self._repeat = False
+        self._media_image_url = None
+        self._volume_level = 0
+        self.update_data(player_data)
+
+    async def async_added_to_hass(self):
+        """Register callback."""
+        self._remove_signal_status = async_dispatcher_connect(
```

Comment:
We can use `self.async_on_remove` to register the remove callback.

https://github.com/home-assistant/core/blob/1776540757544981ddb3457a6166681777c3a163/homeassistant/helpers/entity.py#L472-L474